### DISCUSSION TITLE
Current user permission inside relationship

### DISCUSSION THREAD

**Discussion by ccchapman** [spatie/laravel-permission#2093] [Q&A]
My `User` model has a relationship `posts`.

```php
public function posts()
{
    return $this->hasMany(Post::class);
}
```

I want to add a conditional based on the current user's permission.

```php
public function posts()
{
    return $this->hasMany(Post::class)
      ->when(
          auth()->user()->can("posts.something"),
          fn ($query) => $query->something(),
      );
}
```

This works fine. However, the user's roles become incorrect later in the request:
`auth()->user()->roles` returns more roles than belongs to the user.

Observations:

1. If the `->can()` in the relation is removed, the issue goes away.
1. If I add an `unsetRelation("roles")` to the relation after calling `can()`,
   the issue goes away.

Am I misusing the Laravel or the package? Is this a bug?

**Comment by erikn69** [comment]
Version of package? version of laravel? php? give the complete information, also make a failing test or a minimal application to reproduce the problem
